### PR TITLE
Add native "remove this semicolon" diagnostics

### LINKED ISSUE #6739: Add native "remove this semicolon" diagnostics
See https://github.com/rust-lang/rust/issues/79741.

We should have this diagnostic built-in.

Related things:

* https://github.com/rust-analyzer/rust-analyzer/blob/a0fa522fdaf25daff6a2a9794214f0e0bedc5c24/crates/assists/src/handlers/wrap_return_type_in_result.rs
* https://github.com/rust-analyzer/rust-analyzer/blob/4c00bdea675d33d578cff1c0d7cc3967303e6653/crates/hir_ty/src/diagnostics.rs#L187-L202

### PR DESCRIPTION
**Pull request by ivan770**
Closes #6739 
demo2

### CODE REVIEW

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 1]
```diff
@@ -78,6 +79,12 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         let body_expr = &body[body.body_expr];
         if let Expr::Block { tail: Some(t), .. } = body_expr {
             self.validate_results_in_tail_expr(body.body_expr, *t, db);
+        } else {
+            if let Expr::Block { statements, .. } = body_expr {
+                if let Some(Statement::Expr(id)) = statements.last() {
+                    self.validate_missing_tail_expr(body.body_expr, *id, db);
+                }
+            }
         }
```
Comment (bjorn3) [inline crates/hir_ty/src/diagnostics/expr.rs:88]:
```suggestion
        if let Expr::Block { statements, tail, .. } = body_expr {
            if let Some(t) = tail {
                self.validate_results_in_tail_expr(body.body_expr, *t, db);
            } else if let Some(Statement::Expr(id)) = statements.last() {
                self.validate_missing_tail_expr(body.body_expr, *id, db);
            }
        }
```

Looks a bit nicer IMO.

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 2]
```diff
@@ -317,6 +324,31 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
...
+
+                if let Ok(source_ptr) = source_map.expr_syntax(possible_tail_id) {
+                    self.sink.push(RemoveThisSemicolon {
+                        file: source_ptr.file_id,
+                        expr: source_ptr.value,
+                    });
+                }
+            }
+        }
+    }
```
Comment (bjorn3) [inline crates/hir_ty/src/diagnostics/expr.rs:351]:
```suggestion
        let possible_tail_ty = if let Some(possible_tail_ty) = self.infer.type_of_expr.get(possible_tail_id) {
            possible_tail_ty
        } else {
            return;
        };

        if mismatch.actual != Ty::unit() || mismatch.expected != *possible_tail_ty {
            return;
        }

        let (_, source_map) = db.body_with_source_map(self.owner.into());
        if let Ok(source_ptr) = source_map.expr_syntax(possible_tail_id) {
            self.sink.push(RemoveThisSemicolon {
                file: source_ptr.file_id,
                expr: source_ptr.value,
            });
        }
    }
```

This significantly reduces the indentation.

File: crates/hir_ty/src/diagnostics/expr.rs [hunk 3]
```diff
@@ -317,6 +322,35 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
...
+            Some(m) => m,
+            None => return,
+        };
+
+        let possible_tail_ty =
+            if let Some(possible_tail_ty) = self.infer.type_of_expr.get(possible_tail_id) {
+                possible_tail_ty
+            } else {
+                return;
+            };
```
[the remark this answers is not in the document]
Comment (matklad) [reply crates/hir_ty/src/diagnostics/expr.rs:342]: @ivan770 super minor, but we usually prefer a `match` over `if let else`, as that's shorter:

```rust
        let possible_tail_ty = match self.infer.type_of_expr.get(possible_tail_id) {
            Some(it) => it,
            None => return,
        };
```

File: crates/ide/src/diagnostics/fixes.rs [hunk 4]
```diff
@@ -102,6 +102,27 @@ impl DiagnosticWithFix for MissingOkInTailExpr {
...
+            .text_range();
+
+        let edit = TextEdit::delete(semicolon);
+        let source_change =
+            SourceFileEdit { file_id: self.file.original_file(sema.db), edit }.into();
+
+        Some(Fix::new("Remove this semicolon", source_change, semicolon))
+    }
+}
+
```
Comment (matklad) [inline crates/ide/src/diagnostics/fixes.rs:125]: Let's add a test here as well (parent module has some)

File: crates/ide/src/diagnostics/fixes.rs [hunk 5]
```diff
@@ -102,6 +102,27 @@ impl DiagnosticWithFix for MissingOkInTailExpr {
     }
 }
 
+impl DiagnosticWithFix for RemoveThisSemicolon {
+    fn fix(&self, sema: &Semantics<RootDatabase>) -> Option<Fix> {
+        let root = sema.db.parse_or_expand(self.file)?;
+
+        let semicolon = self
+            .expr
```
Comment (matklad) [inline crates/ide/src/diagnostics/fixes.rs:110]: this would become just

```suggestion
            .expr.semicolon_token()?
```

if we store ExprStmt

**Comment by matklad** [comment]
Looks perfect now, thanks!

bors r+

**Pull request merged**